torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -1892,8 +1896,31 @@ def call_compiled_backward():
                         not backward_state_indices
                     ), "BackwardState requires CompiledAutograd"
                     ctx.maybe_clear_saved_tensors()
+
+                    saved_tensors_use_once = not (
+                        torch.is_grad_enabled() or ctx.is_retain_graph()
+                    )
+
                     if CompiledFunction.compiled_bw is None:
                         assert lazy_backward_info is not None
+
+                        if not saved_tensors_use_once:
+                            fw_metadata.bw_donated_idxs = []
+                            # Update bw_donated_idxs if using lazy_backward_info from `aot_dispatch_autograd`
+                            if (
+                                hasattr(lazy_backward_info, "saved_context")
```

Comment:
hmmm, it's not clear to me that this will properly save this runtime value into the cache.

cc @jamesjwu (this impacts the lazy backward caching behavior): the context is that:

(1) at compile time, AOTAutograd will (unconditionally) compute a `List[int]` of `bw_donated_idxs`, corresponding two which indices of the backward graph inputs can be marked "donated buffers" for inductor

(2) at runtime when the backward is actually invoked and we compiled the backward, if the user specified `.backward(retain_graph=True)`, we need to set this list to be empty before compiling the backward (indicating that these buffers are not actually safe to donate, since autograd is keeping the graph around).

if a user specifies `.backward(retain_graph=True)` and it turns out that we have already compiled a version of this graph that **already** committed to donating the buffers (either because we had a backward cache hit, or we re-ran multiple times in process, or we hit the current code-path where we compile the bw ahead of time), then we want to error.

But it seems to me that in the warm cache case, we need to make sure we save the "new" `fw_metadata. bw_donated_idxs` that we set here.

I forgot how the `fw_metadata` actually gets saved to the cache in the backward case (since I don't see it passed as a value to `try_save_cache_entry()` below). Or maybe the current state is just that we don't support lazy backward caching yet? (in which case, we can kick this problem until then)

Reply:
I don't really think this field belongs in `fw_metadata`, since it's quite literally `backward_metadata`. But the easiest way I see to do this is:

1. Add as an argument to try_save_cache_entry a boolean `saved_tensors_use_once`

2. Modify try_save_cache_entry to set `fw_metadata.bw_donated_idxs` to `[]` if `saved_tensors_used_once` is True. 
3. Pass in saved_tensors_use_once to try_save_cache_entry on line 1940 below. 

That resolves the issue, but assumes that no other compile time wrapper needed the old version of `bw_donated_idxs`. The best solution would be to move it out of fw_metadata, and instead add it as an argument to post_compile, and then do the same steps, but keeping it outside of fw_metadata itself. 

That is, add it as an argument here: 
https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1466
As well as here:
https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/autograd_cache.py#L287
And thread it through the cache. But that's a more complicated change, so if you want to leave it in fw_metadata for now and I can refactor it for the cache, I'd be happy to give it a shot. I think we should at least do the steps above though for correctness.